sympy/core/expr.py

```diff
@@ -1162,17 +1162,23 @@ def as_terms(self):
         for term in Add.make_args(self):
             coeff, _term = term.as_coeff_Mul()
 
-            coeff = complex(coeff)
+            coeff = [S.One, S.Zero]
+            def _coeff(c, d):
+                a, b = coeff
+                coeff[0] = a*c - d*b
+                coeff[1] = a*d + b*c
             cpart, ncpart = {}, []
 
             if _term is not S.One:
                 for factor in Mul.make_args(_term):
                     if factor.is_number:
-                        try:
-                            coeff *= complex(factor)
-                        except (TypeError, ValueError):
-                            pass
-                        else:
+                        if factor.is_real is not None:
```

Comment:
Assumptions should still not be checked. This will cause numerical evaluation.

Reply:
It *might* use numerical evaluation:
```python
>>> Integral(sin(exp(x)), (x, 1, oo)).is_real
>>>
```
If evaluation is undesireable *ever* then the next question is to see what level of change we are willing to tolerate. e.g. without evalulation is this ok?
```python
>>> 3+(2)*I
3 + 2*I
>>> 3+sqrt(2)*I
sqrt(2)*I + 3
```

Reply:
We can do some good while eliminating significant harm in the behavior by limiting the use of evaluation. Conversion to evauation-free handling in core will need an updated means of putting args in order for printing (and that is going to take much more work than this).

Let's see what breaks with this minor adjustment.